Updated Dynamicmap tutorial and new Stream tutorial

jlstevens opened the pull request:
This PR updates the DynamicMap tutorial. Note I'm making very granular commits so I would recommend a squash type merge for this PR.

philippjfr commented:
>so I would recommend a squash type merge for this PR.

Let's start doing that for all PRs unless commits are already factored out into nice chunks.

jlstevens commented:
> Let's start doing that for all PRs unless commits are already factored out into nice chunks.

I normally put some effort into making commits meaningful so I'm not sure we should *always* squash. I recommend it here as I know these small commits aren't worth recording.

jlstevens commented:
As we are planning on a squash commit, I'm happy to show any necessary code fixes in this PR. We can either leave them in this PR or split them out before squashing and merging.

@philippjfr What do you think about the changes in 008c824? It seems to be working the way I want now...

philippjfr commented on holoviews/core/dimension.py:
```diff
@@ -87,6 +87,22 @@ def replace_dimensions(cls, dimensions, overrides):
+                        dropped.append(key)
+                for drop in dropped:
+                    del dmap.data[drop]
+        return dmap.data
```
I'd invert these loops, instead of iterating over the keys once for each kdim this should be iterating over the keys once and filtering each key. Something like

```python
filtered = []
for key, value in dmap.data.items():
    skip = False
    for kd, v in zip(dmap.kdims, key):
          if kdim.values and v not in kdim.values:
               skip = True
     if not skip:
         filtered.append((k, v))
return filtered
```

jlstevens replied:
Sure. I'll implement that now.

jlstevens replied:
Addressed in #1351

jlstevens commented:
I've pulling out the code changes and will be pushing them to #1351. I've now pushed forced a single commit that is just about updating the tutorial. We shouldn't need to do a squash merge now.

philippjfr commented:
I just realized that we've not really documented the lazy nature of the methods on DynamicMap very well. I've not read in detail but I know you've introduced redim but is that ever used in the context of being able to change the dimensions on the elements returned by the DynamicMap? This tutorial is getting pretty long but it would be good to at least introduce the idea that select, relabel, redim etc. also work on the data returned by the DynamicMap.

jlstevens commented:
The tutorial is now much shorter than it was but I do want to resist making it long again. Do you have a short example of the sort of thing would you like to show/discuss?

jlstevens commented:
I've added the new Stream tutorial to this PR (and updated the title). The content is closely enough linked with the ``DynamicMap`` tutorial that I think it is worth keeping them together.

jlstevens commented:
Note: the Streams notebook relies on PR #1356.

philippjfr commented on doc/Tutorials/Streams.ipynb:
```diff
@@ -0,0 +1,826 @@
+    "\n",
+    "In this tutorial we will explore the HoloViews streams system that allows *any* sort of value to be supplied from *anywhere*. This system opens a huge set of new possible visualization types, including continuously updating plots that reflect live data as well as dynamic visualizations that can be interacted with directly, as described in the [LINKED STREAMS] tutorial.\n",
+    "\n",
+    "<center><div class=\"alert alert-info\" role=\"alert\">To use visualize and use a <b>DynamicMap</b> you need to be running a live Jupyter server.<br>This tutorial assumes that it will be run in a live notebook environment.<br>\n",
```
>To use visualize and use a DynamicMap

Small fix needed here.

philippjfr commented on doc/Tutorials/Streams.ipynb:
```diff
@@ -0,0 +1,826 @@
+    "\n",
+    "In this tutorial we will explore the HoloViews streams system that allows *any* sort of value to be supplied from *anywhere*. This system opens a huge set of new possible visualization types, including continuously updating plots that reflect live data as well as dynamic visualizations that can be interacted with directly, as described in the [Linked Streams](Linked_Streams.ipynb) tutorial.\ ...
+    "\n",
+    "<center><div class=\"alert alert-info\" role=\"alert\">To use visualize and use a <b>DynamicMap</b> you need to be running a live Jupyter server.<br>This tutorial assumes that it will be run in a live notebook environment.<br>\n",
```
>To use visualize and use a <b>DynamicMap</b>

Small fix needed here.

jlstevens replied:
Well spotted. Fixed in 3673b90d

philippjfr commented on doc/Tutorials/Dynamic_Map.ipynb:
```diff
@@ -75,46 +72,46 @@
    "metadata": {},
    "source": [
-    "We will now create the ``DynamicMap`` equivalent of the ``HoloMap`` introduced in the [Containers Tutorial](Containers.ipynb#HoloMap). The ``HoloMap`` in that tutorial consisted of ``Image`` elements containing sine ring arrays as defined by the ``sine_array`` function:"
+    "We will now create  ``DynamicMap`` similar to the  ``HoloMap`` introduced in the [Containers Tutorial](Containers.ipynb#HoloMap). The ``HoloMap`` in that tutorial consisted of ``Image`` elements defined by a function returning NumPy arrays called ``sine_array``. Here we will define a ``waves`` function that returns an array pattern parameterized by arbitrary ``alpha`` and ``beta`` parameters ...
```
>We will now create  ``DynamicMap``

a DynamicMap?

jlstevens commented:
I don't think this should hold up an initial merge but I have this note of something I didn't discuss but would like to mention:

* DynamicMap tutorial: unbounded property and slicing restrictions

philippjfr commented:
Looks good, we can polish it more later.

jlstevens commented:
Thanks!
